File: sdks/ruby/lib/blok/logging/logger.rb
# frozen_string_literal: true

module Blok
  module Logging
    # Logger captures structured log entries during node execution.
    # Thread-safe via Mutex. Entries can be retrieved for inclusion
    # in ExecutionResult.logs.
    class Logger
      # @param min_level [Symbol] Minimum log level to capture (default: :info)
      def initialize(min_level = LogLevel::INFO)
        @min_level = min_level
        @entries   = []
        @mutex     = Mutex.new
      end

      # Log a debug message.
      # @param message [String] Log message
      # @param fields [Hash, nil] Optional structured fields
      def debug(message, fields: nil)
        log(LogLevel::DEBUG, message, fields)
      end

      # Log an info message.
      # @param message [String] Log message
      # @param fields [Hash, nil] Optional structured fields
      def info(message, fields: nil)
        log(LogLevel::INFO, message, fields)
      end

      # Log a warning message.
      # @param message [String] Log message
      # @param fields [Hash, nil] Optional structured fields
      def warn(message, fields: nil)
        log(LogLevel::WARN, message, fields)
      end

      # Log an error message.
      # @param message [String] Log message
      # @param fields [Hash, nil] Optional structured fields
      def error(message, fields: nil)
        log(LogLevel::ERROR, message, fields)
      end

      # Return all captured log entries.
      # @return [Array<LogEntry>]
      def entries
        @mutex.synchronize { @entries.dup }
      end

      # Return log entries as formatted strings for ExecutionResult.logs.
      # @return [Array<String>]
      def lines
        @mutex.synchronize { @entries.map(&:to_s) }
      end

      # Clear all captured entries.
      def clear
        @mutex.synchronize { @entries.clear }
      end

      private

      def log(level, message, fields)
        return if LogLevel.priority(level) < LogLevel.priority(@min_level)

        entry = LogEntry.new(level: level, message: message, fields: fields)
        @mutex.synchronize { @entries << entry }
      end
    end
  end
end

File: sdks/csharp/src/Blok.Core/Logging/Logger.cs
namespace Blok.Core.Logging;

/// <summary>
/// Logger captures log entries for inclusion in ExecutionResult.logs.
/// Thread-safe via lock.
/// </summary>
public class Logger
{
    private readonly List<LogEntry> _entries = new();
    private readonly object _lock = new();
    private readonly LogLevel _minLevel;

    public Logger(LogLevel minLevel = LogLevel.Info)
    {
        _minLevel = minLevel;
    }

    private void Log(LogLevel level, string message, Dictionary<string, object?>? fields = null)
    {
        if (level < _minLevel) return;

        var entry = new LogEntry
        {
            Level = level,
            Message = message,
            Timestamp = DateTime.UtcNow,
            Fields = fields
        };

        lock (_lock)
        {
            _entries.Add(entry);
        }
    }

    /// <summary>Log a debug message.</summary>
    public void Debug(string message) => Log(LogLevel.Debug, message);

    /// <summary>Log a debug message with fields.</summary>
    public void Debug(string message, Dictionary<string, object?> fields) => Log(LogLevel.Debug, message, fields);

    /// <summary>Log an info message.</summary>
    public void Info(string message) => Log(LogLevel.Info, message);

    /// <summary>Log an info message with fields.</summary>
    public void Info(string message, Dictionary<string, object?> fields) => Log(LogLevel.Info, message, fields);

    /// <summary>Log a warning message.</summary>
    public void Warn(string message) => Log(LogLevel.Warn, message);

    /// <summary>Log a warning message with fields.</summary>
    public void Warn(string message, Dictionary<string, object?> fields) => Log(LogLevel.Warn, message, fields);

    /// <summary>Log an error message.</summary>
    public void Error(string message) => Log(LogLevel.Error, message);

    /// <summary>Log an error message with fields.</summary>
    public void Error(string message, Dictionary<string, object?> fields) => Log(LogLevel.Error, message, fields);

    /// <summary>
    /// Get all captured log entries.
    /// </summary>
    public List<LogEntry> Entries()
    {
        lock (_lock)
        {
            return new List<LogEntry>(_entries);
        }
    }

    /// <summary>
    /// Get log entries as formatted strings for ExecutionResult.logs.
    /// </summary>
    public List<string> Lines()
    {
        lock (_lock)
        {
            return _entries.Select(e => e.ToString()).ToList();
        }
    }

    /// <summary>
    /// Clear all captured entries.
    /// </summary>
    public void Clear()
    {
        lock (_lock)
        {
            _entries.Clear();
        }
    }
}
